packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -0,0 +1,219 @@
+    batch: ExecutionBatch,
+    state: ExecutionEngineState
+  ) {
+    return Promise.all(
+      batch.map((entry) => this._executeBatchEntry(entry, state))
+    );
+  }
+
+  private async _executeBatchEntry(
+    future: Future,
+    state: ExecutionEngineState
+  ): Promise<ExecutionResult> {
+    let current: JournalableMessage = this._startOrRestartFor(future, state);
+    await this._apply(state, current);
+
+    const context = {
+      executionState: state.executionStateMap[future.id],
+    };
+
+    const exectionStrategy = state.strategy.executeStrategy(context);
+
+    while (!isExecutionResult(current)) {
+      context.executionState = state.executionStateMap[future.id];
+
+      if (isOnChainAction(current)) {
+        current = await state.transactionService.onchain(current);
+      } else if (isOnchainResult(current)) {
+        current = (await exectionStrategy.next(current)).value;
+      } else {
+        current = (await exectionStrategy.next(null)).value;
+      }
+
+      await this._apply(state, current);
+    }
+
+    return current;
+  }
+
+  private _startOrRestartFor(
+    future: Future,
+    state: ExecutionEngineState
+  ): FutureStart | FutureRestart {
+    const executionState = state.executionStateMap[future.id];
+
+    if (executionState === undefined) {
+      return this._initCommandFor(future, state);
+    }
+
+    return { type: "execution-restart", futureId: future.id };
+  }
+
+  private async _apply(
+    state: ExecutionEngineState,
+    message: JournalableMessage
+  ): Promise<void> {
+    await state.journal.record(message);
+    state.executionStateMap = executionStateReducer(
+      state.executionStateMap,
+      message
+    );
+  }
+
+  private _resolveAddress(
+    potential: string | AccountRuntimeValue | undefined,
+    { accounts }: { accounts: string[] }
+  ) {
+    if (typeof potential === "string") {
+      return potential;
+    }
+
+    if (
+      isRuntimeValue(potential) &&
+      potential.type === RuntimeValueType.ACCOUNT
+    ) {
+      return accounts[potential.accountIndex];
+    }
+
+    if (potential === undefined) {
+      return accounts[0];
+    }
+
+    throw new IgnitionError(
+      `Unable to resolve address: ${safeToString(potential)} `
+    );
+  }
+
+  private _initCommandFor(
+    future: Future,
+    { accounts }: { executionStateMap: ExecutionStateMap; accounts: string[] }
+  ): FutureStart {
+    const strategy = "basic";
+
+    switch (future.type) {
+      case FutureType.ARTIFACT_CONTRACT_DEPLOYMENT:
+      case FutureType.NAMED_CONTRACT_DEPLOYMENT:
+        const state: FutureStart = {
+          type: "execution-start",
+          futureId: future.id,
+          futureType: future.type,
+          strategy,
+          // status: ExecutionStatus.STARTED,
+          dependencies: [...future.dependencies].map((f) => f.id),
+          // history: [],
+          storedArtifactPath: "./artifact.json",
+          storedBuildInfoPath: "./build-info.json",
```

Comment:
These should be created here

Reply:
This will be done as part of https://github.com/NomicFoundation/ignition/issues/109